### PR TITLE
Overhaul link and emphasis resolution

### LINKED ISSUE #280: Emphasis surrounding link with asterisk in the label
Take the following example:

``` 
*test [some*](link)*
````

most implementations, including the official commonmark dingus, seem to parse this "correctly" to

```
<p><em>test <a href="link">some*</a></em></p>
```

This library fails at it:

```diff
diff --git a/test/original/emphasis_and_strong.unit b/test/original/emphasis_and_strong.unit
index 13a6dd2..1df522d 100644
--- a/test/original/emphasis_and_strong.unit
+++ b/test/original/emphasis_and_strong.unit
@@ -81,3 +81,7 @@ a_b_ a__b__ a*b* a**b**
 _a_b c_d_ __a__b c__d__ *a*b c*d* **a**b c**d**
 <<<
 <p><em>a_b c_d</em> <strong>a__b c__d</strong> <em>a</em>b c<em>d</em> <strong>a</strong>b c<strong>d</strong></p>
+>>> link with asterisk
+*test [some*](link)*
+<<<
+<p><em>test <a href="link">some*</a></em></p>
```
```
INPUT:
'''r
*test·[some*](link)*
'''            
           
EXPECTED:
'''r
<p><em>test·<a·href="link">some*</a></em></p>
'''

GOT:
'''r
<p><em>test·[some</em>](link)*</p>
'''
``` 

I looked a bit at it and frankly I'm not sure how to correctly handle this with the current architecture.

### LINKED ISSUE #318: Incorrect Emphasis Parsing Of Interior Delimiter Runs That Are Multiples Of 3
The markdown package incorrectly parses interior closing and opening delimiter runs that are _both_ multiples of 3.

Both the [CommonMark](https://spec.commonmark.org/0.29/#example-415) and [GitHub Flavored Markdown](https://github.github.com/gfm/#example-425) specifications note that "_When the lengths of the interior closing and opening delimiter runs are both multiples of 3, though, they can match to create emphasis_"  (click on the document links to go to the examples in the specifications). The example from both specs show the example Markdown input data:

```
foo***bar***baz

foo******bar*********baz
 ```
should produce the following HTML output, respectively:
```
<p>foo<em><strong>bar</strong></em>baz</p>

<p>foo<strong><strong><strong>bar</strong></strong></strong>***baz</p>
```
However, the markdown package does not properly parse these emphasis tags and processes them as ordinary text. The following sample code:
```
import 'package:markdown/markdown.dart';

void main() {
  print(markdownToHtml('foo***bar***baz'));
  print(markdownToHtml('foo******bar*********baz'));
}
```
Produces the following incorrect HTML output:
```
<p>foo***bar***baz</p>

<p>foo******bar*********baz</p>
```

### LINKED ISSUE #327: Incorrect Parsing of Escaped Unbalanced Parentheses in Link Destination
The markdown package incorrectly parses link destinations with escaped unbalanced parentheses.

Both the [CommonMark](https://spec.commonmark.org/0.29/#example-494) and [GitHub Flavored Markdown](https://github.github.com/gfm/#example-506) specifications note that unbalanced parentheses in a link destination are allowed if they are escaped  (click on the document links to go to the examples in the specifications). The example from both specs show the example Markdown input data:

```
[link](foo\(and\(bar\))
 ```
should produce the following HTML output:
```
<p><a href="foo(and(bar)">link</a></p>
```
However, the markdown package does not properly parse this link. The following sample code:
```
import 'package:markdown/markdown.dart';

void main() {
  print(markdownToHtml('[link](foo\(and\(bar\))'));
}
```
Produces the following incorrect HTML output:
```
<p>[link](foo(and(bar))</p>
```

### LINKED ISSUE #328: Incorrect Parsing of Link Destination with Escaped Symbols
The markdown package incorrectly parses link destinations with escaped symbols.

Both the [CommonMark](https://spec.commonmark.org/0.29/#example-496) and [GitHub Flavored Markdown](https://github.github.com/gfm/#example-508) specifications note that "_Parentheses and other symbols can also be escaped, as usual in Markdown_"  (click on the document links to go to the examples in the specifications). The example from both specs show the example Markdown input data:

```
[link](foo\)\:)
 ```
should produce the following HTML output:
```
<p><a href="foo):">link</a></p>
```
However, the markdown package does not properly parse this link. The following sample code:
```
import 'package:markdown/markdown.dart';

void main() {
  print(markdownToHtml('[link](foo\)\:)'));
}
```
Produces the following incorrect HTML output:
```
<p><a href="foo">link</a>:)</p>
```

### LINKED ISSUE #329: Incorrect Parsing of Link Title Containing Backslash Escape, Entity and Numeric Character
The markdown package incorrectly parses link titles containing backslash escape, entity, and numeric character references.

Both the [CommonMark](https://spec.commonmark.org/0.29/#example-502) and [GitHub Flavored Markdown](https://github.github.com/gfm/#example-514) specifications note that "_Backslash escapes and entity and numeric character references may be used in titles_"  (click on the document links to go to the examples in the specifications). The example from both specs show the example Markdown input data:

```
[link](/url "title \"&quot;")
 ```
should produce the following HTML output:
```
<p><a href="/url" title="title &quot;&quot;">link</a></p>
```
However, the markdown package does not properly parse this link. The following sample code:
```
import 'package:markdown/markdown.dart';

void main() {
  print(markdownToHtml('[link](/url "title \"&quot;")'));
}
```
Produces the following incorrect HTML output:
```
<p>[link](/url "title "&quot;")</p>
```

### LINKED ISSUE #330: Incorrect Parsing of Link Grouping Precedence Over Emphasis Grouping
The markdown package incorrectly parses a link with emphasis not giving link grouping precedence over emphasis grouping.

Both the [CommonMark](https://spec.commonmark.org/0.29/#example-517) and [GitHub Flavored Markdown](https://github.github.com/gfm/#example-529) specifications note that "_...illustrate the precedence of link text grouping over emphasis grouping_"  (click on the document links to go to the examples in the specifications). The example from both specs show the example Markdown input data:

```
*[foo*](/uri)
 ```
should produce the following HTML output:
```
<p>*<a href="/uri">foo*</a></p>
```
However, the markdown package does not properly parse this link. The following sample code:
```
import 'package:markdown/markdown.dart';

void main() {
  print(markdownToHtml('*[foo*](/uri)'));
}
```
Produces the following incorrect HTML output:
```
<p><em>[foo</em>](/uri)</p>
```

### LINKED ISSUE #331: Incorrect Parsing of Reference Link with Escaped Square Bracket
The markdown package incorrectly parses reference links that contain escaped brackets in the link text.

Both the [CommonMark](https://spec.commonmark.org/0.29/#example-525) and [GitHub Flavored Markdown](https://github.github.com/gfm/#example-537) specifications note that "_The link text may contain balanced brackets, but not unbalanced ones, unless they are escaped_"  (click on the document links to go to the examples in the specifications). The example from both specs show the example Markdown input data:

```
[link \[bar][ref]

[ref]: /uri
 ```
should produce the following HTML output:
```
<p><a href="/uri">link [bar</a></p>
```
However, the markdown package does not properly parse this reference link. The following sample code:
```
import 'package:markdown/markdown.dart';

void main() {
  print(markdownToHtml('[link \[bar][ref]\n\n[ref]: /uri'));
}
```
Produces the following incorrect HTML output:
```
<p>[link <a href="/uri">bar</a></p>
```

### LINKED ISSUE #332: Incorrect Parsing of Reference Link Grouping Precedence Over Emphasis Grouping
The markdown package incorrectly parses a reference link with emphasis not giving link grouping precedence over emphasis grouping. **NOTE: This could be a duplicate or at least related to issue #330.**

Both the [CommonMark](https://spec.commonmark.org/0.29/#example-530) and [GitHub Flavored Markdown](https://github.github.com/gfm/#example-542) specifications note that "_...illustrate the precedence of link text grouping over emphasis grouping_"  (click on the document links to go to the examples in the specifications). The example from both specs show the example Markdown input data:

```
*[foo*][ref]

[ref]: /uri
 ```
should produce the following HTML output:
```
<p>*<a href="/uri">foo*</a></p>
```
However, the markdown package does not properly parse this reference link. The following sample code:
```
import 'package:markdown/markdown.dart';

void main() {
  print(markdownToHtml('*[foo*][ref]\n\n[ref]: /uri'));
}
```
Produces the following incorrect HTML output:
```
<p><em>[foo</em>]<a href="/uri">ref</a></p>
```

### LINKED ISSUE #338: Incorrect Parsing of Shortcut Reference Link Grouping Precedence Over Emphasis Grouping
The markdown package incorrectly parses a shortcut reference link with emphasis not giving shortcut reference link grouping precedence over emphasis grouping. **NOTE: This could be a duplicate or at least related to issues #330 and #332.**

Both the [CommonMark](https://spec.commonmark.org/0.29/#example-560) and [GitHub Flavored Markdown](https://github.github.com/gfm/#example-572) specifications note that "_...this is a link, because a link label ends with the first following closing bracket._"  (click on the document links to go to the examples in the specifications). The example from both specs show the example Markdown input data:

```
[foo*]: /url

*[foo*]
 ```
should produce the following HTML output:
```
<p>*<a href="/url">foo*</a></p>
```
However, the markdown package does not properly parse this shortcut reference link. The following sample code:
```
import 'package:markdown/markdown.dart';

void main() {
  print(markdownToHtml('[foo*]: /url\n\n*[foo*]'));
}
```
Produces the following incorrect HTML output:
```
<p><em>[foo</em>]</p>
```

### PR DESCRIPTION
**Pull request by srawlins**
Resolution of complex link and emphasis text follows very specific rules which
were incompatible with the current TagState stack. The new algorithms follow
the process outlined in the [CommonMark
spec](https://spec.commonmark.org/0.29/#an-algorithm-for-parsing-nested-emphasis-and-links).

Fixes #280. Fixes #318. Fixes #327. Fixes #328. Fixes #329. Fixes #330. Fixes #331. Fixes #332. Fixes #338.

The crux of the issue which required such an overhaul is that the current
TagState stack did not include any ability to wait to parse a tag's inner text
until it was known that a tag could be closed at the current position, then
parse that inner text, then close the tag. This unfortunately requires a
breaking change for downstream packages which subclass TagSyntax.

* BREAKING: TagSyntax constructor no longer takes an `end` parameter. TagSyntax
  no longer implements `onMatchEnd`. Instead, TagSyntax implements a method
  called 'close' which creates and returns a Node, if a Node can be created and
  closed at the current position. If the TagSyntax instance cannot create a
  Node at the current position, the method should return `null`. Some TagSyntax
  subclasses will unconditionally create a tag in `close`, while others may be
  unable to, such as LinkSyntax, if an inline or reference link could not be
  resolved.

* Loosely, the stack of TagStates is replaced with a stack of Delimiters and a
  tree of parsed HTML nodes.
* Emphasis and strong emphasis, link and image open delimiters are handled with
  the "look for link or image" and "process emphasis" algorithms.
* We combine adjacent text in a more intentional way, and likely more efficient
  manner.
* The _DelimiterRun class is replaced with three classes: abstract Delimiter
  and subclasses SimpleDelimiter, and DelimiterRun.

These changes result in no new spec failures. Emphasis compliance rises from
96% to 99%. Link compliance rises from 90% to 93%. Total CommonMark compliance
rises from 93% to 94%. Total GFM compliance rises from 92% to 93%.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -2,3 +2,4 @@
 .packages
 .pub
 pubspec.lock
+.vscode
```
Comment (kevmoo) [inline .gitignore:5]: Should be in ~/.gitignore-global IMHO

https://dart.dev/guides/libraries/private-files

**Review by kevmoo — COMMENTED** [review]
Woo hoo!

Do we need to do a breaking change here? Or are you not worried about folks who may be using this changed classes?

I'm not really in a position to do a full review of the code. Everything **looks** fine

File: .gitignore (hunk 1 above)
Comment (srawlins) [reply .gitignore:5]: Ah, thanks!

**Comment by srawlins** [comment]
Yeah this would be a breaking change release. I actually think I will land after a breaking null safety:

* markdown 4.0.0 -> null safety.
* markdown 5.0.0 -> this change.

**Comment by kevmoo** [comment]
> Yeah this would be a breaking change release. I actually think I will land after a breaking null safety:
> 
> * markdown 4.0.0 -> null safety.
> * markdown 5.0.0 -> this change.

Worth doing them together? Breaking changes are tough enough to deal with...

**Comment by srawlins** [comment]
I could do them together too if that's recommended. I thought it would be better to not block devs on their Null Safety migration on an unrelated breaking API change.

**Comment by kevmoo** [comment]
> I could do them together too if that's recommended. I thought it would be better to not block devs on their Null Safety migration on an unrelated breaking API change.

I don't have a solid sense here. Will the API break hit most folks migrating? Will it be hard to migrate to the new API in `TagSyntax`? Any harder than migrating to null safety?

**Comment by srawlins** [comment]
> I don't have a solid sense here. Will the API break hit most folks migrating? Will it be hard to migrate to the new API in TagSyntax? Any harder than migrating to null safety?

Hmm, it will not hit most users. I'm actually not certain it will really affect anyone... haha. If someone defines a new emphasis syntax, like we have StrikethroughSyntax for GitHub. It would be non-trivial to migrate; just have to sit down and see the new API and figure out out. Maybe not as hard as migrating to null-safety.

But these answers lead me to think it should be fine then to do one breaking release. Very very few users will be affected by the breaking API change.

**Comment by kevmoo** [comment]
The bigger pain is having two breaks to migrate the world through. Better
to consolidate.

**Review by jcollins-g — COMMENTED** [review]
I am not sure if I'm familiar enough with the package and its other users to know, but this looks good from a general code perspective and how dartdoc uses it.   A reading of dartdoc's markdown package usage suggests we wouldn't be impacted by the breaking change.

**Comment by srawlins** [comment]
Yeah... finding someone who is familiar with markdown or this package is pretty tricky. Bob is the original author, but he hasn't touched it in years... hmm de hmm.

**Review by kevmoo — APPROVED** [review]
I trust you, @srawlins 😄

**Comment by mjordan56** [comment]
@srawlins can you confirm for me which release of the markdown package contains this fix? I'm assisting @domesticmouse with the null safety migration of the flutter_markdown package. I just tried verifying the fixes for the bug issues I wrote against the markdown package and all of them failed. The issues had not been fixed and the output from the markdown package was the same. We're using version ^4.0.0-nullsafety.0.

**Comment by srawlins** [comment]
Can you show me an example. I just got it to work for me:

```none
$ mkdir test_pkg
$ cd test_pkg
$ cat <<EOF >pubspec.yaml
name: test_pkg
version: 0.0.1
environment:
  sdk: '>=2.12.0-0 <3.0.0'

dependencies:
  markdown: 4.0.0-nullsafety.0
EOF
$ ~/code/dart-sdk/sdk/xcodebuild/ReleaseX64/dart-sdk/bin/dart --version
Dart SDK version: 2.12.0-edge.2c41e14b8be53d35d88449999dacef2bfaef2aac (be) (Mon Jan 25 16:07:03 2021 +0000) on "macos_x64"
$ ~/code/dart-sdk/sdk/xcodebuild/ReleaseX64/dart-sdk/bin/pub get
Resolving dependencies... 
+ args 1.6.0
+ charcode 1.2.0-nullsafety.3
+ markdown 4.0.0-nullsafety.0
+ meta 1.3.0-nullsafety.6
Changed 4 dependencies!
$ ~/code/dart-sdk/sdk/xcodebuild/ReleaseX64/dart-sdk/bin/pub run markdown
Precompiling executable... 
Precompiled markdown:markdown.
[link](<foo\>)
<p>[link](&lt;foo&gt;)</p>

$ cat <<EOF >a.dart
void main() {
  print(markdownToHtml('[link](<foo\\>)'));
}
EOF
$ ~/code/dart-sdk/sdk/xcodebuild/ReleaseX64/dart-sdk/bin/dart a.dart
<p>[link](&lt;foo&gt;)</p>

$
```

(Note the single backslash when sending the Markdown via stdin, and the two backslashes when writing a Dart String in a Dart program.)

**Comment by mjordan56** [comment]
Hi, @srawlins. Thank you for the quick response. I appreciate you verifying that the fix for the link and emphasis issues are in version 4.0.0-nullsafety.0 build. I'm taking a closer look at the flutter_markdown test cases that are failing. So far, I can confirm you are correct in that the case you highlighted, and many of the cases I reported, are indeed fixed. So far, I'm finding the problem with the failing tests to be either in the flutter_markdown test code, failing to account for Dart escaping backslashes, or in how the flutter_markdown builder is interpreting the AST. I'll let you know if any of the failures are due to a problem in the markdown package.

Thank you for your assistance, and I apologize for any confusion I may have caused.

cc: @domesticmouse

**Comment by srawlins** [comment]
No problem! Let me know if there's something I missed.

**Pull request merged**